File: homework1/src/warmup.js
// This is homework 1 written by John Hardy and Jordan Sanders

const crypto = require('crypto');

function change(amount) {
  if (amount < 0) {
    throw new RangeError('amount cannot be negative');
  }
  const result = [];
  let remaining = amount;
  [25, 10, 5, 1].forEach((coin) => {
    result.push(Math.floor(remaining / coin));
    remaining %= coin;
  });
  return result;
}

function stripQuotes(arg) {
  let newString = arg;
  newString = newString.replace(/"/g, '');
  newString = newString.replace(/\\/g, '');
  newString = newString.replace(/'/g, '');
  return newString;
}

function scramble(arg) {
  if (arg === '') {
    return '';
  }
  const randomElement = Math.floor(Math.random() * arg.length);
  let string = arg.slice(0, randomElement) + arg.slice(randomElement + 1, arg.length);
  string = arg.charAt(randomElement) + scramble(string);
  return string;
}

function powers(base, limit, callback) {
  let value = 0;
  let exp = 0;
  let nextvalue = 0;
  while (value < limit && nextvalue <= limit) {
    value = Math.pow(base, exp);
    callback(value);
    nextvalue = Math.pow(base, (exp + 1));
    exp += 1;
  }
}

function* powersGenerator(base, limit) {
  let currentPower = 0;
  let currentValue = Math.pow(base, currentPower);
  while (currentValue <= limit) {
    yield currentValue;
    currentPower += 1;
    currentValue = Math.pow(base, currentPower);
  }
}

function say(string) {
  if (string === undefined) {
    return '';
  }
  return string;
}

function interleave(...args) {
  const firstArray = args.shift();
  const secondArray = args;
  if (firstArray.length === 0) {
    return secondArray;
  } else if (secondArray.length === 0) {
    return firstArray;
  }
  let result = [];
  let current = 0;
  function hasNext(val, arr) {
    return val + 1 < arr.length;
  }
  do {
    result.push(firstArray.shift());
    result.push(secondArray.shift());
    current += 1;
  } while (hasNext(current, firstArray) && hasNext(current, secondArray));
  result = result.concat(firstArray);
  result = result.concat(secondArray);
  return result;
}

function cylinder(obj) {
  let { radius = 1, height = 1 } = obj;

  const surfaceArea = () =>
    ((2 * (Math.PI) * radius * height) + (2 * (Math.PI) * (radius * radius)));

  const volume = () => ((Math.PI) * (radius * radius) * height);

  const widen = (factor) => {
    radius *= factor;
  };
  const stretch = (factor) => {
    height *= factor;
  };
  const toString = () => `Cylinder with radius ${radius} and height ${height}`;

  return Object.freeze({
    get radius() {
      return radius;
    },
    get height() {
      return height;
    },
    surfaceArea,
    volume,
    widen,
    stretch,
    toString,
  });
}

function makeCryptoFunctions(k, alg) {
  function encrypt(val) {
    const key = crypto.createCipher(alg, k);
    let result = key.update(val, 'utf8', 'hex');
    result += key.update.final('hex');
    return result;
  }

  function decrypt(val) {
    const key = crypto.createDecipher(alg, k);
    let result = key.update(val, 'hex', 'utf8');
    result += key.update.final('utf8');
    return result;
  }

  return [encrypt, decrypt];
}

module.exports = {
  change,
  stripQuotes,
  scramble,
  powers,
  powersGenerator,
  say,
  interleave,
  cylinder,
  makeCryptoFunctions,
};

File: homework2/warmup.py
"""Homework assignment 2 written by John Hardy and Jordan Sanders."""

import random
from Crypto.Cipher import AES
import requests


def change(cents_left):
    """Calculate change in US currency given a number of cents."""
    if cents_left < 0:
        raise ValueError("amount cannot be negative")
    denominations = [25, 10, 5, 1]
    for d in range(len(denominations)):
        current_denomination = denominations[d]
        denominations[d] = cents_left // current_denomination
        cents_left %= current_denomination
    return tuple(denominations)


def strip_quotes(s):
    """Remove apostrophes and double quotes from a given string."""
    s = s.replace("'", "")
    return s.replace('"', '')


def scramble(string):
    """Randomly permute a given string."""
    result = list(string)
    for x in range(len(result)):
        random_element = random.randrange(len(string))
        result[x], result[random_element] = result[random_element], result[x]
    return "".join(result)


def powers():
    """Yield succesive powers of a base from one to a specific limit."""
    pass


def triples(limit):
    """Return a list of all pythagorean triples up to and including a limit."""
    return [(a, b, c)
            for a in range(1, (limit + 1))
            for b in range(1, (limit + 1))
            for c in range(1, (limit + 1))
            if a**2 + b**2 == c**2 and a < b and b < c]


def say(start_word=None):
    """Chainable function that returns a word based on chained arguments."""
    result = []

    def say_again(word=None):
        if word is None:
            return " ".join(result)
        result.append(word)
        return say_again
    return say_again(start_word)


def interleave(*arguments):
    """Interleave first argument array with remaining arguments."""
    first_array = arguments[0]
    largest = max(len(first_array), len(arguments) - 1)
    result = []
    for i in range(largest):
        if i < len(first_array):
            result.append(first_array[i])
        if i < len(arguments) - 1:
            result.append(arguments[i + 1])
    return result


def Cylinder():
    """Create a Python class for a Cylinder."""
    pass


def make_crypto_functions(key, vector):
    """Use an AES algorithmn to encrypt or decrypt a vector given a key."""
    new_key, new_vector = bytes(key, "utf-8"), bytes(vector, "utf-8")

    def encrypt(to_encrypt):
        encryption_suite = AES.new(new_key, AES.MODE_CBC, new_vector)
        return encryption_suite.encrypt(to_encrypt)

    def decrypt(to_decrypt):
        decryption_suite = AES.new(new_key, AES.MODE_CBC, new_vector)
        return decryption_suite.decrypt(to_decrypt)
    return (encrypt, decrypt)


def random_name(gender, region):
    """Return a random name from the uinames API."""
    pay_load = {"gender": gender, "region": region, "amount": 1}
    r = requests.get("http://uinames.com/api/", params=pay_load)
    response = r.text
    if response == '{"error":"Invalid gender"}':
        raise ValueError(response)
    surname = ""
    for i in range(response.find('rname":"') + 8, response.find('","g')):
        surname += response[i]
    name = ""
    for i in range(response.find('"name":"') + 8, response.find('","s')):
        name += response[i]
    return surname + ", " + name
